### PR TITLE
Can get a DDF by sampling a Spark DDF by size

### PR DESCRIPTION
**Pull request by nhanitvn**
### Description and related tickets, documents
Also:
- The resulted DDF has exact 'size' number of rows
- Sample by fraction now return exact number of rows as specified by fraction
- The old sample by fraction is renamed to indicate that the result is approximated.

Reviewers: @Huandao0812 @phvu @hai-adatao 

### Breaking changes & backward compatible issues
No breaking changes in API. However, we need to take care about performance.

### How to test
Describe how this PR is tested. In case manual testing is required, describe how to do so.

### PR Progress
Make sure all checkboxes below are checked before merged
- [x] Branch is in format `prefix/description` (see [this](http://www.guyroutledge.co.uk/blog/git-branch-naming-conventions/))
- [ ] Merge check has no conflicts. PR checks passed.
- [ ] Code review is done. 

Change-Id: I853d48575247d9179936d2a8c67ae79035ccb959

### CODE REVIEW

File: core/src/main/java/io/ddf/content/IHandleViews.java [hunk 1]
```diff
@@ -24,12 +24,14 @@
    */
   public List<Object[]> getRandomSample(int numSamples, boolean withReplacement, int seed);
 
-  public DDF getRandomSampleByNum(int numSamples,
+  public DDF getRandomSampleByNum(long numSamples,
```
Comment (phvu) [inline core/src/main/java/io/ddf/content/IHandleViews.java]: Does this break lineage history?

Comment (Huandao0812) [inline core/src/main/java/io/ddf/content/IHandleViews.java]: we don't have lineage for sample operation anyway

File: spark/src/main/scala/io/ddf/spark/content/ViewHandler.scala [hunk 2]
```diff
@@ -51,10 +56,50 @@ class ViewHandler(mDDF: DDF) extends io.ddf.content.ViewHandler(mDDF) with IHand
...
+
+      // Get an RDD of rows, create a new SparkDF and then a DDF
+      val sampledRDD: RDD[Row] = joinedRDD.map(x => x._2._2.get)
+      val df: DataFrame = mDDF.getRepresentationHandler.get(classOf[DataFrame]).asInstanceOf[DataFrame]
+      val sampledDF = sqlContext.createDataFrame(sampledRDD, df.schema)
+
+
+      val manager = this.getManager
+      val schema = SchemaHandler.getSchemaFromDataFrame(sampledDF)
+      schema.setTableName(mDDF.getSchemaHandler.newTableName())
```
Comment (Huandao0812) [inline spark/src/main/scala/io/ddf/spark/content/ViewHandler.scala]: you don't need to do this, also we have function newDDFFromSparkDataFrame in SparkDDFManager to create DDF from DataFrame

File: spark/src/main/scala/io/ddf/spark/content/ViewHandler.scala [hunk 3]
```diff
@@ -51,10 +56,50 @@ class ViewHandler(mDDF: DDF) extends io.ddf.content.ViewHandler(mDDF) with IHand
...
+      val sampleDDF = manager.newDDF(manager, sampledDF, Array
+      (classOf[DataFrame]), manager.getNamespace,
+        null, schema)
+
+      mLog.info(">>>>>>> adding ddf to DDFManager " + sampleDDF.getName)
+      this.getDDF.getSchemaHandler.getColumns.foreach{
```
Comment (Huandao0812) [inline spark/src/main/scala/io/ddf/spark/content/ViewHandler.scala]: we have copyFactor function in MetaDataHandler

File: spark/src/main/scala/io/ddf/spark/content/ViewHandler.scala [hunk 4]
```diff
@@ -51,10 +56,50 @@ class ViewHandler(mDDF: DDF) extends io.ddf.content.ViewHandler(mDDF) with IHand
...
+
+    if (numSamples < 0) {
+      throw new IllegalArgumentException("Number of samples must be larger than or equal to 0")
+    }
+
+    if (!withReplacement){
+      mDDF.getSqlHandler.sql2ddf(s"select * from ${mDDF.getSchema.getTableName} order by rand($seed) limit $numSamples")
+    } else {
+      // First, create a Spark DF that contain random numbers from 0 to num_rows - 1
+      val numRows = mDDF.getNumRows
```
Comment (phvu) [inline spark/src/main/scala/io/ddf/spark/content/ViewHandler.scala]: Another way to do this is
```
rddRow.sample(true, 2.0*numSamples/rddRow.count(), seed).zipWithIndex().filter(_._2 < numSamples).map(_._1)
```
i.e. we use Spark's API to sample twice what we need, then only pick the first `numSamples` rows.
Not sure if it is more efficient though.

File: core/src/main/java/io/ddf/content/IHandleViews.java (hunk 1 above)
Comment (phvu) [inline core/src/main/java/io/ddf/content/IHandleViews.java]: then maybe let's change the function names? Those names sound a bit too verbose. Probably something like this:

- `getRandomSample` -> `sample(double fraction, boolean replacement, int seed)`
- `getRandomSampleByNum` -> `sample(long size, boolean replacement, int seed)`
- `getRandomSampleApprox` -> `sampleApprox(double fraction, boolean replacement, int seed)`

Comment (nhanitvn) [inline core/src/main/java/io/ddf/content/IHandleViews.java]: good point. How do you think? @Huandao0812

File: spark/src/main/scala/io/ddf/spark/content/ViewHandler.scala [hunk 5]
```diff
@@ -51,10 +56,50 @@ class ViewHandler(mDDF: DDF) extends io.ddf.content.ViewHandler(mDDF) with IHand
...
+
+
+      val manager = this.getManager
+      val schema = SchemaHandler.getSchemaFromDataFrame(sampledDF)
+      schema.setTableName(mDDF.getSchemaHandler.newTableName())
+      val sampleDDF = manager.newDDF(manager, sampledDF, Array
+      (classOf[DataFrame]), manager.getNamespace,
+        null, schema)
+
+      mLog.info(">>>>>>> adding ddf to DDFManager " + sampleDDF.getName)
```
Comment (Huandao0812) [inline spark/src/main/scala/io/ddf/spark/content/ViewHandler.scala]: this is not necessary, also sampleDDF  does not have name here, only uuid

File: spark/src/main/scala/io/ddf/spark/content/ViewHandler.scala (hunk 4 above)
Comment (nhanitvn) [inline spark/src/main/scala/io/ddf/spark/content/ViewHandler.scala]: Wow, great idea. It avoids join and the param numPartitions in generating randomRDD and thus has high probability to be more efficient than my approach. I think that would just erase my work, lol.

File: core/src/main/java/io/ddf/content/IHandleViews.java (hunk 1 above)
Comment (nhanitvn) [inline core/src/main/java/io/ddf/content/IHandleViews.java]: Resolved, I changed to API names :D

**Comment by nhanitvn** [comment]
The change on API names leads to a lot of changes in ddf-jdbc and PE. So I will close this ticket and open another one focusing on just the logic, not name changes.

**Comment by nhanitvn** [comment]
Here it is https://github.com/ddf-project/DDF/pull/328

**Pull request closed without merging**